devtools/maven/src/main/java/io/quarkus/maven/NativeImageMojo.java

```diff
@@ -125,111 +141,249 @@
-
-                        @Override
-                        public Path getAppClassesDir() {
-                            return classesDir;
-                        }
-
-                        @Override
-                        public Path getTransformedClassesDir() {
-                            // not relevant for this mojo
-                            throw new UnsupportedOperationException();
-                        }
-
-                        @Override
-                        public Path getWiringClassesDir() {
-                            // not relevant for this mojo
-                            throw new UnsupportedOperationException();
-                        }
-
-                        @Override
-                        public Path getConfigDir() {
-                            return classesDir;
-                        }
-
-                        @Override
-                        public Map<Path, Set<String>> getTransformedClassesByJar() {
-                            return Collections.emptyMap();
-                        }
-                    })
-                    .pushOutcome(RunnerJarOutcome.class, new RunnerJarOutcome() {
-                        final Path runnerJar = buildDir.toPath().resolve(finalName + "-runner.jar");
-                        final Path originalJar = buildDir.toPath().resolve(finalName + ".jar");
-
-                        @Override
-                        public Path getRunnerJar() {
-                            return runnerJar;
-                        }
-
-                        @Override
-                        public Path getLibDir() {
-                            return runnerJar.getParent().resolve("lib");
-                        }
+        final AppCreator.Builder creatorBuilder = AppCreator.builder();
+        CurateOutcome curateOutcome = null;
+        AugmentOutcome augmentOutcome = null;
+        RunnerJarOutcome runnerJarOutcome = null;
 
-                        @Override
-                        public Path getOriginalJar() {
-                            return originalJar;
+        if (!buildDir.isDirectory() || !new File(buildDir, "lib").isDirectory()) {
+            // The runner JAR has not been built yet, so we are going to build it
+
+            final AppArtifact appCoords;
+            DefaultArtifact appMvnArtifact = null;
+            if (appArtifact == null) {
+                appMvnArtifact = new DefaultArtifact(project.getArtifact().getGroupId(),
+                        project.getArtifact().getArtifactId(),
+                        project.getArtifact().getClassifier(),
+                        project.getArtifact().getArtifactHandler().getExtension(),
+                        project.getArtifact().getVersion());
+                appCoords = new AppArtifact(appMvnArtifact.getGroupId(), appMvnArtifact.getArtifactId(),
+                        appMvnArtifact.getClassifier(), appMvnArtifact.getExtension(),
+                        appMvnArtifact.getVersion());
```

Comment:
So I see why we support the capability to point to another jar but I don't see why we need to support this case? Could you explain it to me?

Thanks!

Reply:
Sorry, I am not sure I understand your question. If the app artifact was not configured explicitly by the user, we assume the current project is the application (as it is today).
The requirement to build the native image for another project is coming from the requirement to integrate external tests into the platform.

Reply:
We discussed it on Zulip. @aloubyansky answered my questions.